matchmaking/matchmaker.py

```diff
@@ -0,0 +1,344 @@
+from datetime import timezone
+from itertools import chain
+
+import networkx as nx
+
+from matchmaking.models import Match, MatchRequest
+
+
+def key(*args):
+    'standardized key used for lookup dicts in this file'
+    return tuple(sorted(args))
+
+
+class MatchMaker:
+    """
+    MatchMaker collects data about who is requesting a match and what matches they have had recently and it returns
+    matches that optimize our scoring function.
+
+    Usage:
+    ```
+    matches = MatchMaker(
+        match_request_since_date=datetime.now().astimezone(timezone.utc) - timedelta(weeks=2)
+    ).run()
+    ```
+
+    If someone has met too recently then we don't allow them to be matched. The score then is:
+     * a base score of 1.0
+     * a per-person boost of up to 1.0 based on how recently they've been matched. We score people higher if they
+       haven't used this feature recently or ever before
+     * a pair-penalty if that same pair of people has met fairly recently.
+
+    The score can be modified without worrying about the rest of the algorithm, the thing to remember is that the
+    algorithm tries to make matches to maximize the total score.
+
+    The formal computer science problem that we're actually solving here is "maximum weight matching". And the algorithm
+    we chose to use for actually making the matches is from the networkx library and runs in O(num_people^3)
+    see: https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.matching.max_weight_matching.html  # noqa
+    At some point, this might be too slow. At that point we can look into other algorithms, such as this approximate
+    algorithm https://web.eecs.umich.edu/~pettie/papers/ApproxMWM-JACM.pdf which supposedly runs in O(num_people).
+
+    The matching algorithm only finds optimal pairs of people, but sometimes this means that people are left out. So
+    following the matching phase, we collect all the unmatched people and attempt to match them with an existing group.
+
+    The final output of `run()` is a list of matches each of which lists the people involved, the score of their match,
+    and the topic selected for them.
+    """
+    def __init__(self, match_request_since_date):
+        """Initialize the MatchMaker
+
+        :param match_request_since_date: only matches for people that requested a match after this date
+        """
+        self.match_request_since_date = match_request_since_date
+
+        self._now = datetime.datetime.now().astimezone(timezone.utc)
+
+        # CONFIG PARAMS
+        # How far to look back in the past for earlier matches. This is used in scoring so that the score is higher if
+        # the people being matched haven't been matched in a while or ever.
+        self.match_since_date = self._now - datetime.timedelta(weeks=26)
+
+        # give a score boost if we haven't seen them in a while or ever
+        self._recent_match_boost_halflife = datetime.timedelta(weeks=6)
```

Comment:
At some point when this is much more mature, that might be reasonable. But for now it's much more convenient to keep the configuration embedded in the code rather than in a separate file.